pmlc/dialect/pxa/transforms/stencil_generic.cc

```diff
@@ -0,0 +1,164 @@
+// Copyright 2020 Intel Corporation
+
+#include "pmlc/dialect/pxa/transforms/stencil_generic.h"
+
+// TODO: Just seeing if the stencil.cc includes work
+#include "mlir/ADT/TypeSwitch.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassOptions.h"
+
+#include "pmlc/dialect/eltwise/ir/ops.h"
+#include "pmlc/dialect/pxa/analysis/strides.h"
+#include "pmlc/dialect/pxa/ir/ops.h"
+#include "pmlc/dialect/pxa/transforms/passes.h"
+#include "pmlc/dialect/tile/ir/ops.h"
+
+#include "pmlc/util/logging.h"
+#include "pmlc/util/util.h"
+
+// // TODO: includes etc
+// #include "mlir/Dialect/Affine/IR/AffineOps.h"
+
+// #include "pmlc/util/logging.h"
+
+namespace pmlc::dialect::pxa {
+
+// TODO: Better to maintain bound_idxs as both set & vector, or just vector?
+//   Or could maintain as map from BlockArg to numeric index (i.e. where it
+//   would have been were it a vector)
+// TODO: Also, should probably at least be a small vector (how small?)
+void StencilGeneric::RecursiveBindIndex(
+    const llvm::SmallVector<mlir::BlockArgument, 8> &bound_idxs,
+    const llvm::SmallVector<mlir::Value, 3> &tensors) {
+  auto curr_idx = bound_idxs.size();
+  if (curr_idx == semantic_idx_count) {
+    // This is a legal binding, append it and we're done with this branch
+    legal_permutations.push_back(
+        TensorAndIndexPermutation(tensors, bound_idxs));
+  } else {
+    for (const auto &block_arg : block_args) {
+      // Don't bind same index twice
+      if (std::find(bound_idxs.begin(), bound_idxs.end(), block_arg) !=
+          bound_idxs.end()) {
+        continue;
+      }
+
+      // Verify the requirements for this index with each tensor are all met
+      bool reqs_met = true;
+      for (size_t i = 0; i < tensors.size(); i++) {
+        try { // TODO: probably don't keep long term
+          if (!requirements.at(std::make_pair(i, curr_idx))(tensors[i],
+                                                            block_arg)) {
+            reqs_met = false;
+            break;
+          }
+        } catch (const std::out_of_range &e) {
+          IVLOG(1, "Error message: " << e.what());
+          IVLOG(1, "Requested key was: " << std::make_pair(i, curr_idx));
+          throw;
+        }
+      }
+      if (!reqs_met) {
+        continue;
+      }
+
+      // If we made it to here, this index has appropriate semantics; bind it
+      // and recurse
+      llvm::SmallVector<mlir::BlockArgument, 8> new_bound_idxs(bound_idxs);
+      new_bound_idxs.push_back(block_arg);
+      RecursiveBindIndex(new_bound_idxs, tensors);
+    }
+  }
+}
+
+void StencilGeneric::DoStenciling() {
+  // Initialization
+  // TODO: Something like?: tensors.clear();
```

Comment:
Since the stencil objects are single use, you shoudn't need to do any 'reseting' of things.